eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/ImpTest.hs

```diff
@@ -26,4 +66,83 @@ instance
   where
   initImpNES = initAlonzoImpNES
   impSatisfyNativeScript = impAllegraSatisfyNativeScript
-  fixupTx = alonzoFixupTx
+  fixupTx = babbageFixupTx
+
+babbageFixupScriptWits ::
+  forall era.
+  ( EraUTxO era
+  , ScriptsNeeded era ~ AlonzoScriptsNeeded era
+  , BabbageEraTxBody era
+  , AlonzoEraTxWits era
+  ) =>
+  Tx era ->
+  ImpTestM era (Tx era)
+babbageFixupScriptWits tx = do
+  utxo <- getUTxO
+  contexts <- impGetPlutusContexts tx
+  let
+    plutusToScript ::
+      forall l.
+      PlutusLanguage l =>
+      Plutus l ->
+      ImpTestM era (Script era)
+    plutusToScript p =
+      case mkPlutusScript @era p of
+        Just x -> pure $ fromPlutusScript x
+        Nothing -> error "Plutus version not supported by era"
+  scriptWits <- forM contexts $ \(_, sh, ScriptTestContext plutus _) ->
+    (sh,) <$> plutusToScript plutus
+  let
+    refInputs = tx ^. bodyTxL . referenceInputsTxBodyL
+    lookupPlutusRefScripts txIn =
+      case txinLookup txIn utxo of
+        Just txOut ->
+          case txOut ^. referenceScriptTxOutL of
+            SJust script | isPlutusScript script -> (hashScript script :)
+            _ -> id
+        Nothing -> error $ "TxIn not found in UTxO: " <> show txIn
+  let
+    plutusRefScripts = Set.foldr' lookupPlutusRefScripts mempty refInputs
+    filterRefScripts script = hashScript script `notElem` plutusRefScripts
+    insertData ::
+      ScriptHash (EraCrypto era) ->
+      Map (DataHash (EraCrypto era)) (Data era) ->
+      ImpTestM era (Map (DataHash (EraCrypto era)) (Data era))
+    shNeeded = getScriptsHashesNeeded $ getScriptsNeeded utxo (tx ^. bodyTxL)
+    insertData sh m = do
+      mbyStc <- getScriptTestContext sh
+      case mbyStc of
+        Just (ScriptTestContext _ (PlutusArgs dat spendDat)) -> do
+          let
+            dataHashes =
+              Map.fromList $
+                (hashData @era (Data dat), Data dat)
+                  : [ (hashData @era (Data sDat), Data sDat)
+                    | Just sDat <- [spendDat]
+                    ]
```

Comment:
I remembered one more simplification:
```suggestion
              fromElems (hashData @era) $
                Data dat : [ Data sDat | Just sDat <- [spendDat]]
```

Which uses this `fromElems`: https://github.com/IntersectMBO/cardano-ledger/blob/60e0dcc2da3824e48c46066d917b0eee08947699/libs/cardano-data/src/Data/MapExtras.hs#L261-L268